[jit] Expose class constant through `attr` and `setattr` in object

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29220 [quant][graphmode] Enable foldbn tests
* **#29219 [jit] Expose class constant through `attr` and `setattr` in object**
* #30787 [jit] Add unsafeRemoveConstant for ClassType
* #31012 [jit] Fix getConstant
* #29218 [jit] Preserve constant from ConcreteModuleType to ClassType

Summary:
We added class constant in previous PRs, this PR allows access to
class constant in the object API

Test Plan:
build/bin/test_jit
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18846851](https://our.internmc.facebook.com/intern/diff/D18846851)

zdevito reviewed (commented):
Similar to my comment earlier, script::Module is a public API, so we should not add methods used only internally on it. After https://github.com/pytorch/pytorch/pull/28828 lands, we can make `.attr` return a constant value from the class if no slot of a similar name exists.

jamesr66a reviewed (approved):
Sure, but better with tests

suo commented on torch/csrc/jit/script/object.h:
```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
 
   bool hasattr(const std::string& name) const {
-    return _ivalue()->type()->findAttributeSlot(name).has_value();
+    return _ivalue()->type()->findAttributeSlot(name).has_value() \
```
You don't need a trailing `\` for multi-line expressions in C++, and it makes clang-format not correctly.

zdevito commented on torch/csrc/jit/script/object.h:
```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
+    if (auto r = _ivalue()->type()->findAttributeSlot(name)) {
+      return _ivalue()->getSlot(*r);
+    }
+    if (auto v = _ivalue()->type()->getConstant(name)) {
```
Why is it called `findAttribute...` but `getConstant`? These names feel inconsistent and surprising here.

jerryzh168 replied:
first one is finding the slot and get the value from object, but second one is getting constant directly from type, I think it's reasonable to be different since we do get ivalue from different places. I can change this to a check as well though

zdevito commented on torch/csrc/jit/script/object.h:
```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
       return _ivalue()->getSlot(*r);
     }
+    if (auto v = _ivalue()->type()->getConstant(name)) {
+      return v;
```
Pretty sure this needs to be `*v` are there tests?

zdevito replied:
I can't find a test, one should be added.

zdevito reviewed (changes requested):
I believe there is a bug along an untested pathway added from this PR.

suo commented on torch/csrc/jit/script/object.h:
```diff
@@ -28,34 +28,60 @@ struct TORCH_API Object {
+      return _ivalue()->getSlot(*r);
+    }
+    if (_ivalue()->type()->hasConstant(name)) {
+      return *_ivalue()->type()->getConstant(name);
```
I agree with the naming consistency comment. Our general convention for retrieving `Foo` is to define three functions: 
- `findFoo()` returns a nullable object (like pointer or `optional`) and the user is responsible for checking the null state
- `getFoo()` is a checked retrieval, which throws an exception if the name is not present.
- `hasFoo()` which is just an existence check.

Can we make `ClassType` conform to this convention? Probably in a separate PR stacked under this one.

jerryzh168 replied:
I think we are conforming to this convention already. I'm following attribute for constant APIs
One problem is getAttribute returns a TypePtr which is nullable, and it returns `nullptr` when it fails: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L1445, but for constant we don't have `nullptr` that's why I'm using `c10::optional<IValue>` here, or should I change it to return a None IValue?

zdevito replied:
Can't getConstant just return a non-nullable IValue or throw an exception? Don't worry about conforming to `getAttribute` it is wrong as well.

jerryzh168 replied:
yeah I'm going to throw exception in both `getConstant` and `getAttribute` when the field is not found, added two PRs in the stack.

suo reviewed (changes requested):
looks mostly fine but we should make the retrieval API consistent with other things. Sorry, should have mentioned this on the original PR implementing the API.

zdevito reviewed (approved):
Looks good, but let's also make the API names consistent in another PR.

kostmo commented:
## CircleCI build failures summary
As of commit adfe232:
* **3/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=adfe232904aa8d02fc6b4c35576e69b1558b77fd).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3919379) binary_linux_libtorch_3_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (1/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72946983))
```
Dec 12 05:46:20 [1555/2993] Building CXX object caffe2/CMakeFiles/inline_container_test.dir/serialize/inline_container_test.cc.o
Dec 12 05:46:21 [1556/2993] Building CXX object caffe2/CMakeFiles/KernelFunction_test.dir/__/aten/src/ATen/core/boxing/KernelFunction_test.cpp.o
Dec 12 05:46:21 [1557/2993] Building CXX object caffe2/CMakeFiles/TensorImpl_test.dir/__/aten/src/ATen/core/TensorImpl_test.cpp.o
Dec 12 05:46:22 [1558/2993] Building CXX object caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state_int8.cc.o
Dec 12 05:46:22 [1559/2993] Building CXX object caffe2/CMakeFiles/common_test.dir/core/common_test.cc.o
Dec 12 05:46:23 [1560/2993] Building CXX object caffe2/CMakeFiles/dlconvertor_test.dir/__/aten/src/ATen/test/dlconvertor_test.cpp.o
Dec 12 05:46:24 [1561/2993] Building CXX object caffe2/CMakeFiles/List_test.dir/__/aten/src/ATen/core/List_test.cpp.o
Dec 12 05:46:24 [1562/2993] Building CXX object caffe2/CMakeFiles/native_test.dir/__/aten/src/ATen/test/native_test.cpp.o
Dec 12 05:46:25 [1563/2993] Building CXX object caffe2/CMakeFiles/NeuralNetTest.dir/core/nomnigraph/tests/NeuralNetTest.cc.o
Dec 12 05:46:26 [1564/2993] Building CXX object caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o
Dec 12 05:46:26 FAILED: caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o 
Dec 12 05:46:26 /usr/bin/c++  -DAT_PARALLEL_OPENMP=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -D_FILE_OFFSET_BITS=64 -Dcaffe2_pybind11_state_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -Icaffe2/aten/src/TH -I../aten/src/TH -Icaffe2/aten/src -I../aten/../third_party/catch/single_include -I../aten/src/ATen/.. -Icaffe2/aten/src/ATen -I../third_party/miniz-2.0.8 -I../caffe2/core/nomnigraph/include -I../caffe2/../torch/csrc/api -I../caffe2/../torch/csrc/api/include -I../c10/.. -Ithird_party/ideep/mkl-dnn/include -I../third_party/ideep/mkl-dnn/src/../include -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /opt/intel/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /opt/conda/include/python2.7 -isystem /opt/conda/lib/python2.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /opt/rocm/hip/include -isystem /include -isystem ../third_party/ideep/mkl-dnn/include -isystem ../third_party/ideep/include -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3  -fPIC   -fvisibility=hidden -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fopenmp -std=gnu++14 -MD -MT caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o -MF caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o.d -o caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o -c ../caffe2/python/pybind_state.cc
Dec 12 05:46:26 In file included from ../torch/csrc/jit/script/module.h:9:0,
Dec 12 05:46:26                  from ../torch/csrc/jit/script/module_python.h:4,
Dec 12 05:46:26                  from ../caffe2/python/pybind_state.cc:37:
Dec 12 05:46:26 ../torch/csrc/jit/script/object.h: In member function ‘void torch::jit::script::Object::setattr(const string&, c10::IValue)’:
Dec 12 05:46:26 ../torch/csrc/jit/script/object.h:43:35: error: ‘struct c10::ClassType’ has no member named ‘findConstantSlot’
Dec 12 05:46:26      } else if (_ivalue()->type()->findConstantSlot(name)) {
Dec 12 05:46:26                                    ^
Dec 12 05:46:26 ../torch/csrc/jit/script/object.h: In member function ‘c10::IValue torch::jit::script::Object::attr(const string&) const’:
Dec 12 05:46:26 ../torch/csrc/jit/script/object.h:61:37: error: ‘struct c10::ClassType’ has no member named ‘findConstantSlot’
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3919374) pytorch_macos_10_13_py3_build (2/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72950119))
```
Dec 11 21:50:16 [1500/2966] Building CXX object caffe2/CMakeFiles/scalar_test.dir/__/aten/src/ATen/test/scalar_test.cpp.o
Dec 11 21:50:16 ../aten/src/ATen/test/scalar_test.cpp:68:10: warning: unused variable 'C' [-Wunused-variable]
Dec 11 21:50:16   auto&& C = at::globalContext();
Dec 11 21:50:16          ^
Dec 11 21:50:16 ../aten/src/ATen/test/scalar_test.cpp:16:16: warning: unused variable 'Float' [-Wunused-const-variable]
Dec 11 21:50:16 constexpr auto Float = ScalarType::Float;
Dec 11 21:50:16                ^
Dec 11 21:50:16 2 warnings generated.
Dec 11 21:50:23 [1501/2966] Building CXX object caffe2/CMakeFiles/ssa_test.dir/onnx/ssa_test.cc.o
Dec 11 21:50:23 [1502/2966] Building CXX object caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o
Dec 11 21:50:23 FAILED: caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o 
Dec 11 21:50:23 /Users/distiller/workspace/clang++  -DAT_PARALLEL_OPENMP=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -D_FILE_OFFSET_BITS=64 -Dcaffe2_pybind11_state_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -Icaffe2/aten/src/TH -I../aten/src/TH -Icaffe2/aten/src -I../aten/../third_party/catch/single_include -I../aten/src/ATen/.. -Icaffe2/aten/src/ATen -I../third_party/miniz-2.0.8 -I../caffe2/core/nomnigraph/include -I../caffe2/../torch/csrc/api -I../caffe2/../torch/csrc/api/include -I../c10/.. -Ithird_party/ideep/mkl-dnn/include -I../third_party/ideep/mkl-dnn/src/../include -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /Users/distiller/workspace/miniconda3/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /Users/distiller/workspace/miniconda3/include/python3.7m -isystem /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /opt/rocm/hip/include -isystem /include -isystem ../third_party/ideep/mkl-dnn/include -isystem ../third_party/ideep/include -Wno-deprecated -fvisibility-inlines-hidden -Wno-deprecated-declarations -Xpreprocessor -fopenmp -I/usr/local/include -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -faligned-new -fno-math-errno -fno-trapping-math -Wno-unused-private-field -Wno-missing-braces -Wno-c++14-extensions -Wno-constexpr-not-const -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3  -isysroot /Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.9 -fPIC   -fvisibility=hidden -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o -MF caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o.d -o caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o -c ../caffe2/python/pybind_state.cc
Dec 11 21:50:23 In file included from ../caffe2/python/pybind_state.cc:26:
Dec 11 21:50:23 In file included from ../caffe2/opt/onnxifi_transformer.h:12:
Dec 11 21:50:23 In file included from ../caffe2/opt/backend_transformer_base.h:5:
Dec 11 21:50:23 In file included from ../caffe2/opt/bound_shape_inferencer.h:4:
Dec 11 21:50:23 ../caffe2/opt/shape_info.h:77:25: warning: comparison of integers of different signs: 'std::__1::vector<caffe2::TensorBoundShape_DimType, std::__1::allocator<caffe2::TensorBoundShape_DimType> >::size_type' (aka 'unsigned long') and 'int' [-Wsign-compare]
Dec 11 21:50:23         dim_type.size() > idx, dim_type.size(), "vs", dim_type.size());
Dec 11 21:50:23         ~~~~~~~~~~~~~~~ ^ ~~~
Dec 11 21:50:23 ../c10/util/Logging.h:96:24: note: expanded from macro 'CAFFE_ENFORCE'
Dec 11 21:50:23     if (C10_UNLIKELY(!(condition))) {                               \
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo commented on torch/csrc/jit/script/object.h:
```diff
@@ -28,33 +28,60 @@ struct TORCH_API Object {
+          v.type()->python_str(),
+          "'");
+      _ivalue()->setSlot(*slot, std::move(v));
+    } else if (_ivalue()->type()->findConstantSlot(name)) {
```
hasConstant more directly expresses intent I think

jerryzh168 committed:
Update on "[jit] Expose class constant through `attr` and `setattr` in object"


Summary:
We added class constant in previous PRs, this PR allows access to
class constant in the object API

Test Plan:
build/bin/test_jit
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18846851](https://our.internmc.facebook.com/intern/diff/D18846851)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Expose class constant through `attr` and `setattr` in object"


Summary:
We added class constant in previous PRs, this PR allows access to
class constant in the object API

Test Plan:
build/bin/test_jit
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18846851](https://our.internmc.facebook.com/intern/diff/D18846851)

[ghstack-poisoned]
